test/vinyl-luatest/gh_6568_replica_initial_join_deletion_of_compacted_run_files_test.lua

```diff
@@ -0,0 +1,82 @@
+local cluster = require('test.luatest_helpers.cluster')
+local fio = require('fio')
+local t = require('luatest')
+
+local g = t.group('gh-6568-replica-initial-join-deletion-of-compacted-run-files')
+
+local s_id = 0
+local master_vinyl_dir = ''
+local replica_vinyl_dir = ''
+g.before_all(function()
+    local helpers = require('test.luatest_helpers')
+
+    g.cluster = cluster:new({})
+
+    local vinyl_memory = 256
+
+    local master_uri = helpers.instance_uri('master')
+    local master_box_cfg = {
+        checkpoint_count = 1,
+        listen           = master_uri,
+        vinyl_memory     = vinyl_memory,
+    }
+    g.master = g.cluster:build_and_add_server({alias = 'master',
+                                               box_cfg = master_box_cfg})
+
+    local replica_box_cfg = {
+        listen       = helpers.instance_uri('replica'),
+        vinyl_memory = vinyl_memory,
+        replication  = {
+            master_uri,
+        },
+        read_only    = true,
+    }
+    g.replica = g.cluster:build_and_add_server({alias   = 'replica',
+                                                box_cfg = replica_box_cfg})
+
+    g.master:start()
+    s_id, master_vinyl_dir = g.master:exec(function()
+        local fio = require('fio')
+
+        local s = box.schema.space.create('test', {engine = 'vinyl'})
+        s:create_index('primary')
+
+        -- unfortunately, we cannot use a closure here
+        for i = 1, 256 do s:insert{i} end
+        box.snapshot()
+
+        return s.id, fio.cwd()
+    end)
+
+    g.replica:start()
+    replica_vinyl_dir = g.replica:exec(function()
+        local fio = require('fio')
+
+        return fio.cwd()
+    end)
+end)
+
+g.after_all(function()
+    g.cluster:drop()
+end)
+
+g.test_replication_compaction_cleanup = function()
+    local run_files_sz = function(vinyl_dir)
+        local index_dir = fio.pathjoin(vinyl_dir, ('%d/0/'):format(s_id))
+        local sz = 0
+        for _, file_name in pairs(fio.listdir(index_dir)) do
+            local file_path = fio.pathjoin(index_dir, file_name)
+            sz = sz + fio.stat(file_path).size
+        end
+        return sz
+    end
+
+    local master_run_files_sz = run_files_sz(master_vinyl_dir)
+    local replica_run_files_sz = run_files_sz(replica_vinyl_dir)
+    local sz_diff = replica_run_files_sz - master_run_files_sz
+    local msg = 'the total size of run files directories of master and' ..
+                'replica must not differ by more than %.1f%%'
+    local perm_rel_sz_diff = 0.5
+    msg = msg:format(perm_rel_sz_diff * 100)
+    t.assert_le(sz_diff / master_run_files_sz, perm_rel_sz_diff, msg)
+end
```

Comment:
The test is flaky:

```
$ yes vinyl-luatest/gh_6568_replica_initial_join_deletion_of_compacted_run_files_test.lua | head -n100 | xargs ./test-run.py
...
Statistics:
* pass: 28
* fail: 1
Failed tasks:
- [vinyl-luatest/gh_6568_replica_initial_join_deletion_of_compacted_run_files_test.lua,
  null]
```

Please fix. Maybe, instead of checking that the number of files on the replica is roughly the same as the number of files on the master, you could just check that all garbage files have been removed? E.g. by comparing `index:stat().run_count` with the number of run files on disk. Then you wouldn't need to limit memory on the master, which would also speed up the test.

Also, 10 seconds is still too long. If you can't reduce the test run time down to a couple of seconds, you should mark it as long (`long_run` in `suite.ini`), but better speed it up even more, of course.

Reply:
Fixed: changed the test log. Now we create a space with large tuples, pass vinyl_memory approximately the size enough to hold one tuple, and fill the space with a number of tuples enough for the disk usage overhead without our patch to prove by comparing the total *.run files size with box.stat.vinyl().disk.data.

The test works 3 seconds on my local machine.